### PR TITLE
Initial refactor of cost functions

### PR DESCRIPTION
**Pull request by jd-lara**
@GabrielKS this is a first pass at a rewrite of all the costs associated with different unit types and the market costs.

### CODE REVIEW

**Comment by GabrielKS** [comment]
Companion branches:

- IS: https://github.com/NREL-Sienna/InfrastructureSystems.jl/tree/jdgk/cost_functions
- PSI: https://github.com/NREL-Sienna/PowerSimulations.jl/tree/jdgk/cost_functions_2

File: src/PowerSystems.jl [hunk 1]
```diff
@@ -334,6 +334,12 @@ export get_data
 export iterate_components
 export get_time_series_multiple
 export get_variable_cost
+export get_no_load_cost
+export get_start_up
+export get_shut_down
+export get_incremental_offer_curves
+export get_decremental_offer_curves
+export get_ancillary_services
```
Comment (jd-lara) [inline src/PowerSystems.jl:343]:
```suggestion
export get_ancillary_service_offers
```

File: src/parsers/power_system_table_data.jl [hunk 2]
```diff
@@ -815,54 +815,38 @@ function calculate_variable_cost(
...
-
     elseif length(var_cost) == 1
-        # if there is only one point, use it to determine the constant $/MW cost
-        var_cost = LinearFunctionData(var_cost[1][1] * fuel_cost + vom)
-        fixed = 0.0
+        # if there is only one point, use it to determine the constant fuel/MW cost
+        var_cost = AverageRateCurve(LinearFunctionData(0, var_cost[1][1]), 0.0)
     end
-    return var_cost, fixed, fuel_cost
+    # @show FuelCurve(var_cost, fuel_cost)  # TODO test correctness of this conversion
```
Comment (jd-lara) [inline src/parsers/power_system_table_data.jl:848]: is this still pending?

File: src/models/cost_functions/MarketBidCost.jl [hunk 3]
```diff
@@ -26,9 +26,17 @@ mutable struct MarketBidCost <: OperationalCost
...
+    }
     "Variable Cost TimeSeriesKey"
-    decremental_offer_curves::Union{Nothing, IS.TimeSeriesKey, PiecewiseLinearData}
+    decremental_offer_curves::Union{
+        Nothing,
+        IS.TimeSeriesKey,
+        CostCurve{InputOutputCurve{PiecewiseLinearData}},
+    }
     "Bids for the ancillary services"
     ancillary_services::Vector{Service}
```
Comment (jd-lara) [inline src/models/cost_functions/MarketBidCost.jl]: for future reference we need to refactor the ancillary services bids too and support price/quantity bids

File: src/models/cost_functions/ValueCurves.jl [hunk 4]
```diff
@@ -1,45 +1,81 @@
...
-# EQUALITY
+# BASE METHODS
 Base.:(==)(a::InputOutputCurve, b::InputOutputCurve) =
     (get_function_data(a) == get_function_data(b))
 
 Base.:(==)(a::T, b::T) where {T <: Union{IncrementalCurve, AverageRateCurve}} =
     (get_function_data(a) == get_function_data(b)) &&
     (get_initial_input(a) == get_initial_input(b))
 
+"Get an `InputOutputCurve` representing `f(x) = 0`"
```
Comment (jd-lara) [inline src/models/cost_functions/ValueCurves.jl:63]: nice touch

**Review by jd-lara — COMMENTED** [review]
The PR looks good to me. Market Bid cost will require some additional effort due to new discovered requirements but this is ok.

File: src/parsers/power_system_table_data.jl (hunk 2 above)
Comment (GabrielKS) [reply src/parsers/power_system_table_data.jl:848]: Yeah and `test_power_system_table_data.jl` is suggesting that I did the math wrong in the other section too

**Comment by GabrielKS** [comment]
All test files pass except `test_power_system_table_data.jl` at the moment, though there's a bit of other cleanup I'd like to do, e.g., restricting things to just use appropriate subtypes of `OperationalCost`

File: src/parsers/power_system_table_data.jl [hunk 5]
```diff
@@ -1047,8 +1052,9 @@ function make_thermal_generator_multistart(
         no_load_cost = 0.0
     else
         (no_load_x, no_load_cost) = first(var_cost)
+        @warn "Strange math occurring here (part 1)"
```
Comment (GabrielKS) [inline src/parsers/power_system_table_data.jl:1055]: @jd-lara, could you check what we're doing with `no_load_cost` here and at `part 2` below? I don't entirely understand why we are subtracting this out in the first place, and I especially don't get why we're subtracting something from power as well as from cost. `part 1` here is a [direct translation of the old logic](https://github.com/NREL-Sienna/PowerSystems.jl/pull/1056/files#diff-25b200340315b2133dda94fd2e46f429b1a28a3ae9c9e63e26813d3c5865e7cbR1052); `part 2` is adding `no_load_cost` to `fixed` because we don't have a `no_load_cost` field in `ThermalGenerationCost`. I think this code is not covered by tests, the test [here](https://github.com/NREL-Sienna/PowerSystems.jl/pull/1071/files#diff-f7ca8d41191dcc468819e4637c4b7e9c23b36ff33054a881572f6d4d8c59cddaR105) doesn't test any `ThermalMultiStart`s.

File: src/parsers/power_system_table_data.jl [hunk 6]
```diff
@@ -1085,7 +1091,13 @@ function make_thermal_generator_multistart(
         shutdown_cost = 0.0
     end
 
-    op_cost = MultiStartCost(var_cost, no_load_cost, fixed, startup_cost, shutdown_cost)
+    @warn "Strange math occurring here (part 2)"
```
Comment (GabrielKS) [inline src/parsers/power_system_table_data.jl:1094]: @jd-lara (`part 2`)

**Comment by GabrielKS** [comment]
All tests now pass.

**Review by jd-lara — COMMENTED** [review]
These changes look good to me. I won't merge yet until we start some exploration of the PSI updates.

**Comment by rodrigomha** [comment]
Do we have methods to unpack the points of PieceWisePointCurves? Or any other InputOutput curve?

We should have a method that wraps:
```
power_cost_curve = PiecewisePointCurve([(1.0, 1.0), (2.0, 3.0), (3.0, 6.0), (4.0, 10.0), (5.0, 15.0)])
input_points, output_points = (zip(get_points(power_cost_curve)...) .|> collect)
```

Also, are we expecting users to use the `get_points` method? I like the presentation of the get_points when you print the curve, so I think it would be good to use this one to print in the REPL
```
julia> get_points(power_cost_curve)
5-element Vector{@NamedTuple{x::Float64, y::Float64}}:
 (x = 1.0, y = 1.0)
 (x = 2.0, y = 3.0)
 (x = 3.0, y = 6.0)
 (x = 4.0, y = 10.0)
 (x = 5.0, y = 15.0)
```

instead of:
```
julia> power_cost_curve
PiecewisePointCurve(PiecewiseLinearData(@NamedTuple{x::Float64, y::Float64}[(x = 1.0, y = 1.0), (x = 2.0, y = 3.0), (x = 3.0, y = 6.0), (x = 4.0, y = 10.0), (x = 5.0, y = 15.0)]))
```

**Comment by GabrielKS** [comment]
@rodrigomha I'm still ironing out the public interface (have taken a few stabs at this but I keep getting the domain-specific terminology wrong) but yes there should ultimately be a way to get all the x-coordinates and a way to get all the y-coordinates. In fact these functions exist [at the underlying FunctionData level](https://github.com/NREL-Sienna/InfrastructureSystems.jl/blob/241514d1b15282f1259f9a780b2c62a8fb4c6238/src/function_data.jl#L95), I just need to figure out how best to wrap them. Definitely [planning](https://github.com/NREL-Sienna/PowerSystems.jl/blob/259b6cbceae49017eaabf495c5fe5c3e959406cc/src/models/cost_functions/cost_aliases.jl#L84) a more friendly `println`/etc. presentation as well.

**Comment by rodrigomha** [comment]
> @rodrigomha I'm still ironing out the public interface (have taken a few stabs at this but I keep getting the domain-specific terminology wrong) but yes there should ultimately be a way to get all the x-coordinates and a way to get all the y-coordinates. In fact these functions exist [at the underlying FunctionData level](https://github.com/NREL-Sienna/InfrastructureSystems.jl/blob/241514d1b15282f1259f9a780b2c62a8fb4c6238/src/function_data.jl#L95), I just need to figure out how best to wrap them. Definitely [planning](https://github.com/NREL-Sienna/PowerSystems.jl/blob/259b6cbceae49017eaabf495c5fe5c3e959406cc/src/models/cost_functions/cost_aliases.jl#L84) a more friendly `println`/etc. presentation as well.

This is good, we can just make a wrapper of `get_x_coords`, in some form (like `get_input_points`) such that we can call it directly to the InputOutputCurve, rather than the `function_data` itself.

Regarding the presentation, we should discuss if using `pretty_tables` or other form makes sense here or not.

**Comment by jd-lara** [comment]
> Do we have methods to unpack the points of PieceWisePointCurves? Or any other InputOutput curve?
> 
> We should have a method that wraps:
> 
> ```
> power_cost_curve = PiecewisePointCurve([(1.0, 1.0), (2.0, 3.0), (3.0, 6.0), (4.0, 10.0), (5.0, 15.0)])
> input_points, output_points = (zip(get_points(power_cost_curve)...) .|> collect)
> ```
> 
> Also, are we expecting users to use the `get_points` method? I like the presentation of the get_points when you print the curve, so I think it would be good to use this one to print in the REPL
> 
> ```
> julia> get_points(power_cost_curve)
> 5-element Vector{@NamedTuple{x::Float64, y::Float64}}:
>  (x = 1.0, y = 1.0)
>  (x = 2.0, y = 3.0)
>  (x = 3.0, y = 6.0)
>  (x = 4.0, y = 10.0)
>  (x = 5.0, y = 15.0)
> ```
> 
> instead of:
> 
> ```
> julia> power_cost_curve
> PiecewisePointCurve(PiecewiseLinearData(@NamedTuple{x::Float64, y::Float64}[(x = 1.0, y = 1.0), (x = 2.0, y = 3.0), (x = 3.0, y = 6.0), (x = 4.0, y = 10.0), (x = 5.0, y = 15.0)]))
> ```

There is an open issue for the printing. We should merge this to continue with the implementation in PSI.

File: src/PowerSystems.jl (hunk 1 above)
Comment (GabrielKS) [reply src/PowerSystems.jl]: Done

File: src/models/cost_functions/MarketBidCost.jl (hunk 3 above)
Comment (GabrielKS) [reply src/models/cost_functions/MarketBidCost.jl]: Done

File: src/parsers/power_system_table_data.jl (hunk 2 above)
Comment (GabrielKS) [reply src/parsers/power_system_table_data.jl]: This is on somewhat firmer ground now but I'm putting it in the TODO issue

File: src/models/cost_functions/variable_cost.jl [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+a [`ValueCurve`][@ref] that may represent input-output, incremental, or average rate data.
+The default units for the x-axis are megawatts and can be specified with `power_units`.
+"""
+@kwdef struct FuelCurve{T <: ValueCurve} <: ProductionVariableCost{T}
+    "The underlying `ValueCurve` representation of this `ProductionVariableCost`"
+    value_curve::T
+    "The units for the x-axis of the curve; defaults to natural units (megawatts)"
+    power_units::UnitSystem = UnitSystem.NATURAL_UNITS
+    "Either a fixed value for fuel cost or the name of a fuel cost time series name"
+    fuel_cost::Union{Float64, String}
```
Comment (GabrielKS) [inline src/models/cost_functions/variable_cost.jl:51]: @daniel-thom Should this be
```suggestion
    fuel_cost::Union{Float64, IS.TimeSeriesKey}
```
?

**Comment by GabrielKS** [comment]
Tracking issue is https://github.com/NREL-Sienna/PowerSystems.jl/issues/1092, I'd love some substantive review of the added source files before I merge this.

File: src/models/cost_functions/variable_cost.jl [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+@kwdef struct CostCurve{T <: ValueCurve} <: ProductionVariableCost{T}
+    "The underlying `ValueCurve` representation of this `ProductionVariableCost`"
+    value_curve::T
+    "The units for the x-axis of the curve; defaults to natural units (megawatts)"
+    power_units::UnitSystem = UnitSystem.NATURAL_UNITS
+end
+
+CostCurve(value_curve) = CostCurve(; value_curve)
+
+Base.:(==)(a::CostCurve, b::CostCurve) =
```
Comment (daniel-thom) [inline src/models/cost_functions/variable_cost.jl:32]: Why is this necessary?

File: src/models/cost_functions/variable_cost.jl [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+`power_units`.
+"""
+@kwdef struct CostCurve{T <: ValueCurve} <: ProductionVariableCost{T}
+    "The underlying `ValueCurve` representation of this `ProductionVariableCost`"
+    value_curve::T
+    "The units for the x-axis of the curve; defaults to natural units (megawatts)"
+    power_units::UnitSystem = UnitSystem.NATURAL_UNITS
+end
+
+CostCurve(value_curve) = CostCurve(; value_curve)
```
Comment (daniel-thom) [inline src/models/cost_functions/variable_cost.jl:30]: Is it painful to implement all of these positional argument constructors? You could require that all cost structs be constructed with kwargs. It might help with consistency.

File: src/models/cost_functions/HydroGenerationCost.jl [hunk 10]
```diff
@@ -0,0 +1,34 @@
+"""
```
Comment (daniel-thom) [inline src/models/cost_functions/HydroGenerationCost.jl:1]: Would it help users to make cost_functions a submodule? Maybe it helps searchability.

File: src/parsers/power_models_data.jl [hunk 11]
```diff
@@ -427,33 +427,41 @@ function make_thermal_gen(gen_name::AbstractString, d::Dict, bus::ACBus, sys_mba
...
             # constant term and make the fixed cost the constant term; in a future update,
-            # there will be no separation between the PolynomialFunctionData and the fixed
+            # there will be no separation between the QuadraticFunctionData and the fixed
             # cost.
             # This transforms [3.0, 1.0, 4.0, 2.0] into [(1, 4.0), (2, 1.0), (3, 3.0)]
             coeffs = enumerate(reverse(d["cost"][1:(end - 1)]))
-            cost = PolynomialFunctionData(Dict((i, c / sys_mbase^i) for (i, c) in coeffs))
+            coeffs = Dict((i, c / sys_mbase^i) for (i, c) in coeffs)
+            quadratic_degrees = [2, 1, 0]
+            (keys(coeffs) <= Set(quadratic_degrees)) || throw(
```
Comment (daniel-thom) [inline src/parsers/power_models_data.jl:451]: Is this idiomatic Julia? It also seems more natural to me to say `if error-condition throw` or `error-condition && throw`. Asserts use this form.

File: src/parsers/power_system_table_data.jl [hunk 12]
```diff
@@ -1267,7 +1280,8 @@ function make_renewable_generator(
     base_power = gen.base_mva
     var_cost, fixed, fuel_cost =
         calculate_variable_cost(data, gen, cost_colnames, base_power)
-    operation_cost = TwoPartCost(var_cost, fixed)
+    @assert fixed == 0 "RenewableGenerationCost cannot have a fixed cost, got $fixed with variable cost $varcost"
```
Comment (daniel-thom) [inline src/parsers/power_system_table_data.jl:1283]:
```suggestion
    @assert fixed == 0 "RenewableGenerationCost cannot have a fixed cost, got $fixed with variable cost $var_cost"
```

File: test/test_constructors.jl [hunk 13]
```diff
@@ -18,10 +18,11 @@
...
 
 @testset "Generation Constructors" begin
-    tThreePartCost = ThreePartCost(nothing)
-    @test tThreePartCost isa IS.InfrastructureSystemsType
-    tTwoPartCost = TwoPartCost(nothing)
-    @test tTwoPartCost isa IS.InfrastructureSystemsType
+    for T in subtypes(PSY.OperationalCost)
+        isabstracttype(T) || (@test T(nothing) isa IS.InfrastructureSystemsType)
+    end
+    # TODO add concrete subtypes of ProductionVariableCost?
```
Comment (daniel-thom) [inline test/test_constructors.jl:24]: Does this need to get done?

File: src/models/cost_functions/MarketBidCost.jl [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+- `decremental_offer_curves::Union{Nothing, IS.TimeSeriesKey, PiecewiseLinearData}`: Buy Offer Curves data, can be a time series or a fixed PiecewiseLinearData
+- `ancillary_service_offers::Vector{Service}`: Bids for the ancillary services
+"""
+@kwdef mutable struct MarketBidCost <: OperationalCost
+    no_load_cost::Float64
+    """start-up cost at different stages of the thermal cycle.
+    Warm is also referred to as intermediate in some markets"""
+    start_up::StartUpStages
+    "shut-down cost"
+    shut_down::Float64
```
Comment (jd-lara) [inline src/models/cost_functions/MarketBidCost.jl:28]: Based on the latest analysis of the curves, these fields should be `Union{Float64, IS.TimeSeriesKey}` to support being read from time series. This we can implement in a later PR after merging this one and time series

Comment (GabrielKS) [reply src/models/cost_functions/MarketBidCost.jl:28]: Added to https://github.com/NREL-Sienna/PowerSystems.jl/pull/1071#discussion_r1579880930

File: src/models/cost_functions/variable_cost.jl (hunk 7 above)
Comment (GabrielKS) [reply src/models/cost_functions/variable_cost.jl:51]: No, keep it String
